### ISSUE TITLE
QgsDataItem::providerKey has different semantic in subclasses

### ISSUE THREAD

**Issue by elpaso** [qgis/qgis3.0_api#114]
`QgsDataItem::providerKey()` is supposed to return the `QgsDataItemProvider` name (key) but all layer items are (correctly, because that value is required for drag and drop mime types) initiated with the name of the `QgsDataProvider` associated with the layer and hence they return the `QgsDataProvider` key and not the `QgsDataItemProvider` key.

I'm proposing to:

- deprecate and rename `QgsDataItem::providerKey` to `QgsDataItem::dataItemProviderKey` (always valued, make the dataItemProviderKey a mandatory argument in `QgsDataItem` ctor)
- add `QgsDataItem::dataProviderKey` (possibly blank because not all data items have a corresponding `QgsDataProvider`)
- rename and add setters accordingly

**Comment by m-kuhn** [comment]
This can be done outside a 4.0 change.

**Comment by elpaso** [comment]
Not with the ctor changes.

**Comment by m-kuhn** [comment]
Agreed. In this case only 90% of this change can be done outside a 4.0 change.
